xgcm/autogenerate.py

```diff
@@ -0,0 +1,272 @@
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+     Dataset with gridinformation used to construct c-grid
+    axes_dims_dict : dict
+     Dict with information on the dimension in ds corrsponding to the xgcm
+     axis. E.g. {'X':'lon','Y':'lat'}
+    axes_coords_dict : dict
+     Dict with information on the coordinates in ds corrsponding to the
+     xgcm axis. E.g. {'X':'geolon','Y':'geolat'}
+    position : {None,tuple, dict}
+     Position of the gridpoints given in 'ds' and the desired position to be
+     generated. Defaults to ('center','left'). Can be a tuple like
+     ('center','left'), or a dict with corresponding axes
+     (e.g. {'X':('center','left'),'Z':('left','center')})
+    wrap : {None, float, dict}
+     Specifies the discontinuity at the boundary to wrap e.g. longitudes
+     without artifacts. Can be defined globally (for all fields defined in
+     axes_dims_dict and axes_coords_dict) {float, None} or per dataset
+     variable (dict e.g. {'longitude':360,'latitude':180})
+    pad : {None, float, 'auto'}
+     Specifies the discontinuity at the boundary to wrap e.g. longitudes
+     without artifacts. Can be defined globally (for all fields defined in
+     axes_dims_dict and axes_coords_dict) {float, None} or per dataset
+     variable ({dict} e.g. {'z':'auto','latitude':0.0})
+    new_name : str
+     Name of the inferred grid variable. Defaults to name+'_inferred'
+    """
+
+    if axes_coords_dict is not None:
+        combo_dict = [axes_dims_dict, axes_coords_dict]
+    else:
+        combo_dict = [axes_dims_dict]
+
+    for di, dd in enumerate(combo_dict):
+        if di == 0:
+            attrs_from_scratch = True
+            infer_dim = False
+        elif di == 1:
+            attrs_from_scratch = False
+            infer_dim = True
+
+        for ax in dd.keys():
+            # Get variable name
+            ax_v = dd[ax]
+            # Get dimension name
+            if infer_dim:
+                ax_d = axes_dims_dict[ax]
+            else:
+                ax_d = ax_v
+
+            # Parse position
+            pos_from, pos_to = parse_position(position, ax)
+            # Pass wrap characteristics
+            is_wrapped = parse_wrap_pad(wrap, ax_v)
+            # Pass pad characteristics
+            is_padded = parse_wrap_pad(pad, ax_v)
+            ds = generate_axis(ds, ax, ax_v, ax_d,
+                               pos_from=pos_from, pos_to=pos_to,
+                               wrap=is_wrapped, pad=is_padded,
+                               new_name=new_name,
+                               attrs_from_scratch=attrs_from_scratch)
+    return ds
+
+
+def parse_wrap_pad(in_val, varname):
+    """Parse wrap or pad parameters"""
+    if isinstance(in_val, dict):
+        try:
+            is_valued = in_val[varname]
+        except KeyError:
+            # Set defaults
+            is_valued = None
+    else:
+        is_valued = in_val
+    return is_valued
```

Comment:
are these private functions? If so, they need to be renamed with a leading underscore (i.e. `_parse_wrap_pad`)